statement.py

```diff
@@ -4,43 +4,396 @@
 from datetime import datetime, timezone, timedelta
 from decimal import Decimal
 from secrets import token_hex
+from itertools import groupby
 
-from trytond.model import ModelView, fields
+from trytond.model import Workflow, ModelView, fields
 from trytond.pool import Pool, PoolMeta
+from trytond.pyson import Eval, Id
+from trytond.rpc import RPC
 from trytond.wizard import (
     Button, StateAction, StateTransition, StateView, Wizard)
 from trytond.transaction import Transaction
 from trytond.config import config
 from .common import get_base_header
 from trytond.i18n import gettext
 from trytond.exceptions import UserError
+from trytond.modules.account_statement.exceptions import (StatementValidateError,
+    StatementValidateWarning)
 
 
-class Origin(metaclass=PoolMeta):
+class Statement(metaclass=PoolMeta):
+    __name__ = 'account.statement'
+
+    @classmethod
+    def cancel(cls, statements):
+        pool = Pool()
+        Origin = pool.get('account.statement.origin')
+
+        origins = [s.origin for s in statements]
+        Origin.cancel(origins)
+
+        super().cancel(statements)
+
+    @classmethod
+    def delete(cls, statements):
+        raise UserError(gettext('account_statement_enable_banking.'
+                'msg_no_allow_delete_only_cancel'))
+
+
+class Line(metaclass=PoolMeta):
+    __name__ = 'account.statement.line'
+
+    @classmethod
+    def cancel_move(cls, lines):
+        pool = Pool()
+        Move = pool.get('account.move')
+        MoveLine = pool.get('account.move.line')
+        Reconciliation = pool.get('account.move.reconciliation')
+        Invoice = pool.get('account.invoice')
+
+        for line in lines:
+            move = line.move
+            if move:
+                to_unreconcile = [x.reconciliation for x in move.lines
+                    if x.reconciliation]
+                if to_unreconcile:
+                    to_unreconcile = Reconciliation.browse([
+                            x.id for x in to_unreconcile])
+                    Reconciliation.delete(to_unreconcile)
+
+                # On possible realted invoices, need to unlink the payment
+                # lines
+                to_unpay = [x for x in move.lines if x.invoice_payment]
+                if to_unpay:
+                    #to_unpay = [MoveLine(x.id) for x in to_unpay]
+                    Invoice.remove_payment_lines(to_unpay)
+
+                cancel_move = move.cancel()
+                cancel_move.origin = line.origin
+                Move.post([cancel_move])
+                mlines = [l for m in [move, cancel_move]
+                    for l in m.lines if l.account.reconcile]
+                if mlines:
+                    MoveLine.reconcile(mlines)
```

Comment:
This method looks a bit naive to me.

I think that at least it should be grouped by account.

Reply:
I think I'm not understanding what you mean. Here we take the lines to reconcile to no send all lines, only which the account has the field setted.